src/slice.rs

```diff
@@ -240,220 +260,342 @@ impl From<RangeFull> for Slice {
-            fn from(r: $index) -> SliceOrIndex {
-                SliceOrIndex::Index(r as isize)
+            fn from(r: $index) -> AxisSliceInfo {
+                AxisSliceInfo::Index(r as isize)
+            }
+        }
+    };
+}
+impl_axissliceinfo_from_index!(isize);
+impl_axissliceinfo_from_index!(usize);
+impl_axissliceinfo_from_index!(i32);
+
+impl From<NewAxis> for AxisSliceInfo {
+    #[inline]
+    fn from(_: NewAxis) -> AxisSliceInfo {
+        AxisSliceInfo::NewAxis
+    }
+}
+
+/// A type that can slice an array of dimension `D`.
+///
+/// This trait is unsafe to implement because the implementation must ensure
+/// that `D`, `Self::OutDim`, `self.in_dim()`, and `self.out_ndim()` are
+/// consistent with the `&[AxisSliceInfo]` returned by `self.as_ref()`.
+pub unsafe trait CanSlice<D: Dimension>: AsRef<[AxisSliceInfo]> {
+    type OutDim: Dimension;
+
+    fn in_ndim(&self) -> usize;
+
+    fn out_ndim(&self) -> usize;
+}
+
+macro_rules! impl_canslice_samedim {
+    ($in_dim:ty) => {
+        unsafe impl<T, Dout> CanSlice<$in_dim> for SliceInfo<T, $in_dim, Dout>
+        where
+            T: AsRef<[AxisSliceInfo]>,
+            Dout: Dimension,
+        {
+            type OutDim = Dout;
+
+            fn in_ndim(&self) -> usize {
+                self.in_ndim()
+            }
+
+            fn out_ndim(&self) -> usize {
+                self.out_ndim()
             }
         }
     };
 }
-impl_sliceorindex_from_index!(isize);
-impl_sliceorindex_from_index!(usize);
-impl_sliceorindex_from_index!(i32);
+impl_canslice_samedim!(Ix0);
+impl_canslice_samedim!(Ix1);
+impl_canslice_samedim!(Ix2);
+impl_canslice_samedim!(Ix3);
+impl_canslice_samedim!(Ix4);
+impl_canslice_samedim!(Ix5);
+impl_canslice_samedim!(Ix6);
+
+unsafe impl<T, Din, Dout> CanSlice<IxDyn> for SliceInfo<T, Din, Dout>
+where
+    T: AsRef<[AxisSliceInfo]>,
+    Din: Dimension,
+    Dout: Dimension,
+{
+    type OutDim = Dout;
+
+    fn in_ndim(&self) -> usize {
+        self.in_ndim()
+    }
+
+    fn out_ndim(&self) -> usize {
+        self.out_ndim()
+    }
+}
 
 /// Represents all of the necessary information to perform a slice.
 ///
-/// The type `T` is typically `[SliceOrIndex; n]`, `[SliceOrIndex]`, or
-/// `Vec<SliceOrIndex>`. The type `D` is the output dimension after calling
-/// [`.slice()`].
+/// The type `T` is typically `[AxisSliceInfo; n]`, `[AxisSliceInfo]`, or
+/// `Vec<AxisSliceInfo>`. The type `Din` is the dimension of the array to be
+/// sliced, and `Dout` is the output dimension after calling [`.slice()`]. Note
+/// that if `Din` is a fixed dimension type (`Ix0`, `Ix1`, `Ix2`, etc.), the
+/// `SliceInfo` instance can still be used to slice an array with dimension
+/// `IxDyn` as long as the number of axes matches.
 ///
 /// [`.slice()`]: struct.ArrayBase.html#method.slice
 #[derive(Debug)]
 #[repr(C)]
-pub struct SliceInfo<T: ?Sized, D: Dimension> {
-    out_dim: PhantomData<D>,
+pub struct SliceInfo<T: ?Sized, Din: Dimension, Dout: Dimension> {
```

Comment:
If we do this, does that actually help? At least we simplify the awful `SliceInfo<_, IxDyn, IxDyn>` invocation

```suggestion
pub struct SliceInfo<T: ?Sized, Din: Dimension, Dout: Dimension = Din> {
```

Reply:
IMO, this change would be too confusing for the small gain in conciseness. I added a comment about the `SliceInfo::<_, IxDyn, IxDyn>` above with a good solution. If we implement `CanSlice<IxDyn>` for `[AxisSliceInfo]`, there really isn't much reason to ever create a `SliceInfo<_, IxDyn, IxDyn>` instance.